### PR TITLE
fix(automation): report the bridge target and resolve the project's bridge port

### PR DESCRIPTION
**Pull request by punal100**
Base: `dev` @ `9ff9d8e8` · Branch: `Dev_Punal_MCP` · Commits: `aa99f0e8` (target in the refusal), `3598730b` (port from project config), `d1da1a87` (closed-set failure reason), `0abc3d63` (port fallback gate + exact-message assertions), `7411bdf1` (dialed client URL + usable-override gate), `0fec239c` (bridge_ack marker precedence) — 6 commits, no unrelated changes

## Summary

Two related gaps made rapid editor open/close cycles and multi-project use painful on the TypeScript stdio surface.

1. Connection failures were uninformative: the bridge layer threw `Automation bridge not connected` (or `Failed to establish connection to Unreal Engine: <cause>`) and the gateway refused with a bare `Unreal Engine is not connected.` — none named the host and ports dialed, so "no listener", "wrong port" and "handshake rejected" were indistinguishable.
2. The client only ever dialed the built-in default port. Projects that set their own `ListenPorts` had to repeat the port in client configuration, and two projects running on the defaults collided: the first editor kept 8090, the second silently lost it, and every client still dialed 8090 — a session could end up driving the wrong project's editor.

The stdio server stays usable without an editor: the single `unreal` tool keeps listing and only the call fails. These commits make that failure name its target and make the target follow the project.

`d1da1a87`, `0abc3d63` and `7411bdf1` are review follow-ups: the lazy-connection race no longer forwards raw exception text to tool output (closed-set reason only, full exception stays in the logs); the project-config fallback is no longer suppressed by an unusable explicit scalar; and the reported target is now the URL the client actually dials (`clientHost`/`clientPort`), not the listen host and port list.

## Changes

**Report the bridge target (`aa99f0e8`)**
- `server/gateway/gateway-execute.ts`: the `NOT_CONNECTED` refusal names the resolved bridge URL when the bridge knows it. `errorCode`, the tool/action echo and the `search` nextCall are unchanged.
- `types/tools/tool-interfaces.ts`: `AutomationRequestBridge` gains an optional `getClientUrl()`; test doubles without it keep the previous wording.
- `automation/bridge-config.ts`: add `formatBridgeTarget` and `bridgeNotConnectedMessage`; the wording keeps the `not connected` marker for transport classification in `services/telemetry-observation.ts`.
- `automation/connection-lifecycle.ts`: accept an optional `describeTarget` and use the shared message for lazy-connection failures.
- `automation/bridge-request-dispatcher.ts`: pass `describeTarget` through and use the shared message for the resolved-but-disconnected path.
- `automation/bridge.ts`: supply `describeTarget` from the resolved config.
- `automation/bridge-request-dispatcher.test.ts`, `tests/unit/tools/unreal-gateway-execute-seam.test.ts`: cover both bridge failure paths (with and without a target) and the gateway refusal with and without a known bridge URL.

**Resolve the port from the project (`3598730b`)**
- `utils/config/ini-reader.ts`: extract the parser and candidate list; add sync `readIniFileSync` and `getProjectSettingSync` for values resolved before the first await. The async reader keeps its exact behaviour.
- `automation/bridge-config.ts`: fall back to the first `ListenPorts` token from `<UE_PROJECT_PATH>/Config/DefaultGame.ini` (section `/Script/McpAutomationBridge.McpAutomationBridgeSettings`, with the same `Saved/Config/<Platform>/Game.ini` fallbacks as the offline project-settings reader) before the built-in 8090. The plugin binds every configured token in order and a busy port drops out of the set, so the first token is the port a client should dial.
- `tests/unit/automation/bridge_config_port_resolution.test.ts`: new coverage.
- Resolution order: `options.port` → `MCP_AUTOMATION_WS_PORT` → `MCP_AUTOMATION_PORT` → project `ListenPorts[0]` → `8090`, with `MCP_AUTOMATION_CLIENT_PORT` overriding the dialed port afterwards. The project fallback applies when none of the scalar sources resolves and no usable port-list source is set: an unusable scalar (`MCP_AUTOMATION_PORT=foo`) falls through to the project value instead of pinning the built-in default. Existing setups and the integration harness keep their exact semantics.

**Map connection failures to a closed-set reason (`d1da1a87`)**
- `automation/bridge-config.ts`: `describeBridgeFailure(cause)` maps a raw exception onto a closed set — `connection refused` / `timed out` / `host unreachable` / `tls failure` / `handshake rejected` / `connection lost` / `server stopped` / `bridge disabled` / `unknown failure`. `bridgeNotConnectedMessage` takes that reason instead of free text.
- `automation/connection-lifecycle.ts`: the full exception stays in the trusted logger (`log.error('Lazy connection failed', error)`); callers see only the mapped reason, so OS or peer-controlled text (for example a handshake response) never reaches tool output.
- `automation/bridge-connection.test.ts`, `tests/unit/automation/bridge_not_connected_reason.test.ts`: per-kind mapping, non-`Error` causes and the "peer text never echoed" case; the `not connected` marker telemetry classifies on is unchanged.
- This path predates the PR — the previous wording also concatenated the cause; the change only removes raw text from caller-visible output. `errorCode`, the tool/action echo and the `search` nextCall are unchanged.

**Tighten the port gate and the failure-message assertions (`0abc3d63`)**
- `automation/bridge-config.ts`: the fallback gate stops covering sources that do not bypass the port chain; an explicit scalar that sanitizes to `null` falls through to the project `ListenPorts` instead of silently pinning `8090`.
- `bridgeNotConnectedMessage` accepts `BridgeFailureReason` instead of `string`, so the closed set is enforced by the type. No caller change.
- `automation/bridge-request-dispatcher.test.ts`: exact caller-visible message assertions (target, reason and trailing hint included) instead of unescaped, non-end-anchored regular expressions.
- `tests/unit/automation/bridge_config_port_resolution.test.ts`: case for an unusable explicit scalar falling back to the project `ListenPorts`.

**Report the dialed URL and gate on usable overrides (`7411bdf1`)**
- `automation/bridge.ts`: `describeTarget` now returns `getClientUrl()`, the `ws://` URL built from `clientHost`/`clientPort` that the socket dials, instead of the listen `host`/`ports` list. The two only differ when they are configured to differ (for example `MCP_AUTOMATION_CLIENT_HOST=::1` + `MCP_AUTOMATION_CLIENT_PORT=8099`). `formatBridgeTarget` is removed as unused, so the gateway refusal and the dispatcher failure now share one `ws://host:port` spelling.
- `automation/bridge-config.ts`: the project `ListenPorts` fallback is gated on usable overrides, not raw presence: a port list must hold a usable token and `options.clientPort` / `MCP_AUTOMATION_CLIENT_PORT` must sanitize. `MCP_AUTOMATION_CLIENT_PORT=not-a-port` or `MCP_AUTOMATION_WS_PORTS=''` now fall through to the project value instead of pinning `8090`; `clientPort` is sanitized like the other scalar sources.
- `automation/bridge-config.ts`: `describeBridgeFailure` classifies structured transport codes before any message text and matches the `bridge_ack` marker before generic words, so peer-influenced handshake text cannot pick the reason (a peer `{"type":"timeout"}` no longer maps to `timed out`); a refused WebSocket upgrade (401/426) maps to `handshake rejected`.
- `automation/bridge-connection.test.ts`: a real `AutomationBridge` with a divergent client host/port asserts the failure names the URL `getClientUrl()` returns. `bridge_config_port_resolution.test.ts` adds unusable `MCP_AUTOMATION_CLIENT_PORT` and unusable port-list cases; `bridge_not_connected_reason.test.ts` adds code-first and `bridge_ack` precedence cases.
- `automation/bridge-config.ts` (`0fec239c`): the `bridge_ack` marker is checked before message-derived transport text, so a peer string holding both `bridge_ack` and `ECONNREFUSED` still maps to `handshake rejected`; classification from a structured `error.code` is unchanged.

Example with the editor closed:

    Unreal Engine is not connected: no bridge listener responded at ws://127.0.0.1:8090.
    errorCode: NOT_CONNECTED | nextCall: { operation=search }

Example resolution with no client-side port configured anywhere (editor running, one client per project):

    project A ListenPorts=8091,8090 -> dials 8091 -> execute control_actor.list succeeds
    project B ListenPorts=8095      -> dials 8095 -> NOT_CONNECTED (nothing listening)
    project C ListenPorts=8090,8091 -> dials 8090 -> default behaviour unchanged

## Related Issues

Related to editor open/close workflows — the server stays usable while the editor is not running; only the call fails, naming the target — and to running several projects against several editor instances at once. With this change a per-project client entry needs nothing but `UE_PROJECT_PATH`.

## Reviewed boundaries

- The native `/mcp` transport builds its own `NOT_CONNECTED` text (`McpNativeGatewayExecuteReceiptBuild.cpp`) and reads its port from project settings, so no mirrored C++ is expected from these commits.
- Connection failures now surface only a closed-set reason plus the dialed URL; the full exception remains in the trusted logs. The target host and ports were already visible by design (loopback-only client).
- `MCP_AUTOMATION_WS_PORTS` remains diagnostics-only; it is not a failover list. The client dials exactly one port.
- Two projects that share the same first `ListenPorts` token still collide by design: the plugin binds in order and the first editor keeps the port.

## Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 📚 Documentation update
- [ ] 🔧 Configuration/build change
- [ ] ♻️ Refactoring (no functional changes)
- [x] 🧪 Test addition/update

## Testing

- [x] Added/updated tests — bridge failure paths with and without a target; gateway refusal with and without a known URL; full caller-visible message assertions; closed-set reason mapping (per-kind, non-`Error`, peer text never echoed, code-first precedence, `bridge_ack` marker, refused upgrades); port resolution across project config, saved per-platform config, `.uproject` paths, explicit env and option overrides, explicit port lists, unusable values, unusable-scalar fallback, unusable client-port and unusable port-list overrides; a real-bridge case asserting the failure names the dialed `getClientUrl()`
- [x] `npm run type-check`
- [x] `npx eslint <changed files> --max-warnings=0`
- [x] `npx vitest run tests/unit/tools/unreal-gateway-execute-seam.test.ts src/automation/bridge-request-dispatcher.test.ts src/automation/bridge-connection.test.ts tests/unit/automation/bridge_config_port_resolution.test.ts tests/unit/automation/bridge_not_connected_reason.test.ts` — 64 passed (5 files)
- [x] Live stdio check in a restarted session, editor closed: tool listed, `execute control_actor.list` → `Unreal Engine is not connected: no bridge listener responded at ws://127.0.0.1:8090.`, `errorCode NOT_CONNECTED`, `nextCall { operation=search }`
- [x] Live stdio check with the editor started (no reconnect, no reload): same call → `success: true`, `worldName: Engine_Test_Cell`, live actors and `liveRevisions`
- [x] Port resolution with no client-side port configured: ini `8091,8090` → dials 8091 (execute succeeds); ini `8095` → dials 8095 and reports `NOT_CONNECTED`; ini `8090,8091` → dials 8090 (unchanged default)
- [ ] Full `npm run test:unit` — 15 failures in this checkout are pre-existing (a real `Saved/MCP/capability-token` plus unrelated contract/normalization cases); the same set reproduces with the PR commits stashed

## Pre-Merge Checklist

- [x] Code follows project style guidelines
- [x] Self-reviewed the code
- [x] Updated relevant documentation (if needed)
- [x] Added/updated tests (if applicable)
- [ ] CI passes

### CODE REVIEW

File: src/automation/bridge-config.ts [hunk 1]
```diff
@@ -45,6 +80,27 @@ export interface AutomationBridgeResolvedConfig {
...
+
+/**
+ * One wording for every "bridge is not there" failure so logs, tool output and
+ * telemetry agree. Callers pass the resolved target and, when available, the
+ * underlying cause. Always includes `not connected` - transport classification
+ * in `services/telemetry-observation.ts` matches that marker.
+ */
+export function bridgeNotConnectedMessage(target?: string, cause?: string): string {
+    const where = target ? ` at ${target}` : '';
+    const why = cause ? `: ${cause}` : '';
```
[the remark this answers is not in the document]
Comment (punal100) [reply src/automation/bridge-config.ts]: Thanks for the detailed chain. Partially valid — mechanism confirmed, severity overstated, and the file reference is off by a layer. Evidence and plan below.

**Confirmed.** The underlying cause can reach tool output: `connection-lifecycle.ts:67-72` does `getErrorMessage(error)` and passes it into `bridgeNotConnectedMessage(target, message)`, which `bridge-config.ts:100` concatenates into the message. This predates this PR — the previous wording was `Failed to establish connection to Unreal Engine: <message>`, so the cause text was already surfaced; these commits only added the dialed target, which is intentional and non-sensitive (the bridge host/port is already logged by design).

**What can actually leak.** OS/peer strings only: `connect ECONNREFUSED 127.0.0.1:8090`, `getaddrinfo ENOTFOUND …`, TLS messages, or text sent back by whatever owns the port. No secrets: the capability token is redacted at the client boundary (`bridge-client.ts:266-281`, `redactPeerText`/`redactPeerError` against the configured token and `redactKnownAutomationCredentials`), the transport is loopback-only by default, and errno text carries no paths or credentials.

**Reachability is narrower than described.**
- `executeAutomationRequest()` (`src/tools/handlers/foundation/dispatch/automation-request-dispatch.ts:76-83`) performs its own preflight and replaces the message with a static string before any bridge call, so it does not forward the raw cause.
- The gateway path is also static: `tryConnectWithBackoff` (`src/unreal-bridge-connection.ts:87-131`) swallows connect errors via `ErrorHandler.retryWithBackoff` and returns a boolean, so `gateway-execute` emits its own `NOT_CONNECTED` refusal with the target only. Verified live with the editor closed: the tool returned `Unreal Engine is not connected: no bridge listener responded at ws://127.0.0.1:8090.` — no errno text.
- The cause text surfaces only through `bridge-request-dispatcher.ts:79`'s `ensureConnected()` throw, i.e. a race where the bridge is seen as connected and drops between the preflight and the send.
- Note the finding cites `bridge-config.ts:100` and `executeAutomationRequest()` together; those are different layers (`bridge-config.ts:100` is only the `cause` concatenation).

**Agreed fix, and we'll fold it into this PR.** Defense in depth is cheap and worth having:
- Map the cause to a closed-set reason for tool output (`connection refused` / `timed out` / `host unreachable` / `tls failure` / `handshake rejected` / `connection lost` / `unknown`), derived from the error `code`/`name` rather than free text.
- Keep the full exception in the trusted logger (`log.error('Lazy connection failed', error)` already does) and in `bridge-state` diagnostics.
- Bound any peer-derived text to a fixed length if it is ever forwarded.
- Tests: closed-set mapping per error kind, and a race-path case asserting the tool message never contains raw exception text.

The `errorCode` contract and the target in the message stay as they are, so no client-visible behaviour changes beyond the reason wording. Native `/mcp` is unaffected: it builds its own `NOT_CONNECTED` text (`McpNativeGatewayExecuteReceiptBuild.cpp`).

Thanks for flagging the injection surface when a foreign process owns the port — that is the one part worth prioritizing regardless of secret exposure.

[the remark this answers is not in the document]
Comment (punal100) [reply src/automation/bridge-config.ts]: Fixed in `d1da1a87` — "fix(automation): map connection failures to a closed-set reason" (third commit on this branch).

What changed:
- `bridge-config.ts`: new `describeBridgeFailure(cause)` maps the raw exception onto a closed set — `connection refused` / `timed out` / `host unreachable` / `tls failure` / `handshake rejected` / `connection lost` / `server stopped` / `bridge disabled` / `unknown failure`. `bridgeNotConnectedMessage` now takes that reason instead of free text.
- `connection-lifecycle.ts`: the full exception stays in the trusted logger (`log.error('Lazy connection failed', error)`); callers see only the mapped reason.
- Tests: per-kind mapping; peer-controlled handshake text is never echoed (`Handshake expected bridge_ack, got {"type":"<script>…"}` → `handshake rejected`); the unknown fallback does not echo paths; closed-set wording. Existing dispatcher/connection expectations moved from the raw `Lazy connection timeout` to `timed out`.

Deliberately unchanged: `errorCode`, the dialed target in the message, and the `not connected` telemetry marker. Native `/mcp` builds its own text (`McpNativeGatewayExecuteReceiptBuild.cpp`), so it is unaffected.

Verification: `npm run type-check`; `eslint … --max-warnings=0`; `npx vitest run` on the touched suites (49 passed: mapping, dispatcher, connection, gateway seam); live stdio check with no listener still returns `Unreal Engine is not connected: no bridge listener responded at ws://127.0.0.1:8090.` with no raw exception text.

File: src/automation/bridge-request-dispatcher.test.ts [hunk 2]
```diff
@@ -323,3 +323,46 @@ describe('AutomationRequestDispatcher send-failure and cancellation edge cases',
...
+    });
+
+    it('stays silent about a target when none is configured', async () => {
+        const { dispatcher } = createDispatcher(new RequestTracker(50), {
+            isConnected: () => false,
+            connectionTimeoutMs: 20,
+        });
+
+        await expect(dispatcher.sendAutomationRequest('get_actor')).rejects.toThrow(
+            /^Automation bridge not connected: timed out/,
```
[the remark this answers is not in the document]
Comment (punal100) [reply src/automation/bridge-request-dispatcher.test.ts]: Both review items applied, plus the nitpick — commit `0abc3d63` ("fix(automation): tighten port fallback and failure-message assertions"), on top of `d1da1a87`:

1. **Scalar fallback (`bridge-config.ts`)** — agreed, that was a real gap. The gate now lists only the sources that bypass the port chain (`options.ports`, `options.clientPort`, `MCP_AUTOMATION_CLIENT_PORT`, `MCP_AUTOMATION_WS_PORTS`), so an explicit scalar that sanitizes to null (`MCP_AUTOMATION_PORT=foo`) falls through to `readProjectListenPort()` instead of pinning 8090. New case: ini `ListenPorts=8092` + `MCP_AUTOMATION_PORT=not-a-port` → dials 8092.
2. **Exact assertions (`bridge-request-dispatcher.test.ts`)** — unanchored regexes replaced with full-message assertions (`toMatchObject({ message })`), including the interpolated target and the trailing hint, so wording and trailing-text regressions now fail.
3. **Nitpick (`BridgeFailureReason`)** — `bridgeNotConnectedMessage(target?, reason?: BridgeFailureReason)` types the closed set directly; no call-site change.

Verification: `npm run type-check`; `eslint … --max-warnings=0`; `npx vitest run` on the touched suites — 80 passed (port resolution, failure-reason mapping, dispatcher, connection, gateway seam).

**Comment by ChiR24** [comment]
Good direction, and the plugin-side assumptions hold — `ListenPorts` really is a comma `FString` (not a `TArray` needing `+` lines), `config=Game` → `DefaultGame.ini`, and `McpConnectionManagerConnection.cpp:31` confirms the first token is what gets bound. Two things I'd fix before this lands; both verified by running the branch.

**1. `bridge.ts:82` — the reported target isn't the dialed one.** `describeTarget` is built from `config.host`/`config.ports`, but the client dials `config.clientHost`/`config.clientPort` (`bridge-client.ts:46`). These agree in the default config, so the tests pass — but they diverge exactly in the multi-project setup this PR targets: with `MCP_AUTOMATION_CLIENT_HOST=::1` + `CLIENT_PORT=8099`, `resolveAutomationBridgeConfig` returns `host=127.0.0.1 ports=[8090]` while the socket goes to `ws://[::1]:8099`, and the error names 127.0.0.1:8090. `getClientUrl()` already exists at `bridge.ts:182` — `describeTarget: () => this.getClientUrl()` fixes it and drops `formatBridgeTarget`, which also collapses the two different target spellings (`ws://host:port` from the gateway vs bare `host:ports` here) into one.

**2. `bridge-config.ts:168` — the bypass gate still pins 8090 on a typo.** `!== undefined` means `MCP_AUTOMATION_WS_PORTS=''` or `MCP_AUTOMATION_CLIENT_PORT=not-a-port` skips `readProjectListenPort()` entirely — with a project on `ListenPorts=9000`, both resolve to `clientPort=8090`. That's the same failure `0abc3d63` fixed for `MCP_AUTOMATION_PORT`, still present for two of the four sources. `sanitizePort(...) !== null` makes it consistent.

Worth a look, lower confidence:

- `bridge-config.ts:122` — the reason is derived from peer-controlled text before any code-based check, so the peer picks its own bucket: a handshake reason of `{"type":"ssl"}` returns `tls failure`, `{"type":"ECONNREFUSED"}` returns `connection refused`. The text isn't echoed any more, but the classification still is. Testing `code` before `message` would close it.
- `bridge-config.ts:124` — `Unexpected server response: 401` / `426` (what `ws` throws when something else owns the port) both fall to `unknown failure`, which is the case the PR most wants to name.
- `.env.example:13` pins `MCP_AUTOMATION_PORT=8091`, so the new fallback is inert for anyone following the documented setup — and 8091 is the second of the plugin's two default tokens.
- `ini-reader.ts:69` returns `Config/Default*.ini` ahead of `Saved/Config/<Platform>/*.ini`, the reverse of UE's layering. `defaultconfig` means it rarely bites, but the new test freezes the order in.

One test note: the dispatcher tests inject `describeTarget` directly, which is why #1 got through — a single case through `new AutomationBridge({...})` asserting the message names the same URL `getClientUrl()` returns would have caught it.

**Comment by punal100** [comment]
@ChiR24 verified both, fixed on `Dev_Punal_MCP` — `7411bdf1`.

**1. The reported target is the dialed one (`src/automation/bridge.ts`)**

`describeTarget` now returns `getClientUrl()` — the `ws://` URL built from `clientHost`/`clientPort` (`bridge-client.ts:46`) — instead of `config.host`/`config.ports`. `formatBridgeTarget` is dropped, so the gateway refusal and the dispatcher failure share one spelling. New regression test builds a real `AutomationBridge({ clientHost: '::1', clientPort: 8099 })` and asserts the message names `bridge.getClientUrl()`.

```diff
-            describeTarget: () => formatBridgeTarget(this.config.host, this.config.ports),
+            describeTarget: () => this.getClientUrl(),
```

**2. The gate now requires a usable override (`src/automation/bridge-config.ts`)**

```diff
-    const hasExplicitPortBypass = options.ports !== undefined
-        || options.clientPort !== undefined
-        || process.env.MCP_AUTOMATION_CLIENT_PORT !== undefined
-        || process.env.MCP_AUTOMATION_WS_PORTS !== undefined;
+    const configuredPortList = options.ports ?? readWsPortsEnv();
+    const hasExplicitPortBypass = configuredPortList.some((value) => sanitizePort(value) !== null)
+        || sanitizePort(options.clientPort) !== null
+        || sanitizePort(process.env.MCP_AUTOMATION_CLIENT_PORT) !== null;
```

```diff
-        clientPort: options.clientPort ?? sanitizePort(process.env.MCP_AUTOMATION_CLIENT_PORT) ?? defaultPort,
+        clientPort: sanitizePort(options.clientPort) ?? sanitizePort(process.env.MCP_AUTOMATION_CLIENT_PORT) ?? defaultPort,
```

`MCP_AUTOMATION_CLIENT_PORT=not-a-port` and `MCP_AUTOMATION_WS_PORTS='not-a-port,'` now fall through to the project `ListenPorts`; both have cases in `bridge_config_port_resolution.test.ts`.

**Lower-confidence item** — structured codes are classified before message text, and the `bridge_ack` marker before generic words, so a peer `{"type":"timeout"}` no longer picks `timed out`; a refused WebSocket upgrade (401/426) maps to `handshake rejected`.

Validation: `tsc --noEmit` clean, eslint `--max-warnings=0` clean, 64 passed across the five touched suites. PR summary updated for the 5 commits.

File: src/automation/bridge-config.ts [hunk 3]
```diff
@@ -45,6 +80,70 @@ export interface AutomationBridgeResolvedConfig {
...
+    if (/(ENOTFOUND|EAI_AGAIN|EHOSTUNREACH|ENETUNREACH|EADDRNOTAVAIL)/.test(codeToken)) return 'host unreachable';
+    if (/(ERR_TLS|CERT|SELF_SIGNED)/.test(codeToken)) return 'tls failure';
+    if (/(ECONNRESET|EPIPE)/.test(codeToken)) return 'connection lost';
+
+    const token = `${code} ${message}`.toUpperCase();
+
+    if (token.includes('ECONNREFUSED')) return 'connection refused';
+    // Our own protocol markers win over generic words a peer can embed in the
+    // received handshake string (for example a `type` value of `timeout`).
+    if (/BRIDGE_ACK/.test(token)) return 'handshake rejected';
```
[the remark this answers is not in the document]
Comment (punal100) [reply src/automation/bridge-config.ts:121]: Valid — fixed in `0fec239c`. The `bridge_ack` marker now decides before the message-derived `ECONNREFUSED` check, so `Handshake expected bridge_ack, got ECONNREFUSED` maps to `handshake rejected`; classification from a structured `error.code` (the trustworthy path) is unchanged. Focused case added to `bridge_not_connected_reason.test.ts`: `tsc --noEmit` clean, eslint clean, 64 passed across the five touched suites.

**Review by ChiR24 — APPROVED** [review]

**Comment by punal100** [comment]
Once Again Thanks for Merging my PR!
Glad to contribute to Open-Source Development
